File: server_express_backup/src/routes/prs.js
import { Router } from "express";
import {
  getAuthenticatedLogin,
  searchReviewRequestedOpenPRs,
  getPR,
  getPRDiff,
  getFileContent,
  normalizePR,
  parsePRUrl,
  repoOwnerFromUrl,
} from "../github.js";
import { readDB, upsertPR, deletePR, computeStatus } from "../db.js";
import { mapWithConcurrency } from "../concurrency.js";

export const router = Router();

function withStatus(pr, db) {
  return { ...pr, status: computeStatus(pr.repo, pr.number, db) };
}

router.get("/prs", async (req, res, next) => {
  try {
    const login = await getAuthenticatedLogin();
    const searchResults = await searchReviewRequestedOpenPRs(login);

    const assignedKeys = new Set();
    await mapWithConcurrency(searchResults, 8, async (item) => {
      const { owner, repo } = repoOwnerFromUrl(item.repository_url);
      if (!owner || !repo) return;
      try {
        const full = await getPR(owner, repo, item.number);
        const pr = normalizePR(full, { owner, repo, source: "assigned" });
        upsertPR(pr);
        assignedKeys.add(`${owner}/${repo}#${item.number}`);
      } catch (err) {
        console.error(`Failed to refresh assigned PR ${owner}/${repo}#${item.number}:`, err.message);
      }
    });

    // Refresh manually-added PRs not covered by the assigned fetch above.
    const dbBefore = readDB();
    const manualEntries = Object.entries(dbBefore.prs).filter(
      ([key, pr]) => pr.source === "manual" && !assignedKeys.has(key)
    );
    await mapWithConcurrency(manualEntries, 8, async ([key, pr]) => {
      const [owner, repo] = pr.repo.split("/");
      try {
        const full = await getPR(owner, repo, pr.number);
        upsertPR(normalizePR(full, { owner, repo, source: "manual" }));
      } catch (err) {
        console.error(`Failed to refresh manual PR ${key}, keeping stale copy:`, err.message);
      }
    });

    // Drop assigned PRs that are no longer open+review-requested (closed, merged, or
    // review no longer needed) — otherwise they'd linger in the db forever since only
    // currently-returned search results get upserted, never pruned.
    for (const [key, pr] of Object.entries(dbBefore.prs)) {
      if (pr.source === "assigned" && !assignedKeys.has(key)) {
        deletePR(pr.repo, pr.number);
      }
    }

    const db = readDB();
    const list = Object.values(db.prs)
      .filter((pr) => pr.state == null || pr.state === "open")
      .map((pr) => withStatus(pr, db))
      .sort((a, b) => new Date(b.updatedAt) - new Date(a.updatedAt));
    res.json(list);
  } catch (err) {
    next(err);
  }
});

router.post("/added-prs", async (req, res, next) => {
  try {
    const { url } = req.body;
    if (!url) {
      return res.status(400).json({ error: "url is required" });
    }
    const { owner, repo, number } = parsePRUrl(url);
    const full = await getPR(owner, repo, number);
    const pr = upsertPR(normalizePR(full, { owner, repo, source: "manual" }));
    res.status(201).json(withStatus(pr, readDB()));
  } catch (err) {
    next(err);
  }
});

router.get("/prs/:owner/:repo/:number/diff", async (req, res, next) => {
  try {
    const { owner, repo, number } = req.params;
    const diff = await getPRDiff(owner, repo, number);
    res.type("text/plain").send(diff);
  } catch (err) {
    next(err);
  }
});

router.get("/prs/:owner/:repo/:number/file", async (req, res, next) => {
  try {
    const { owner, repo, number } = req.params;
    const { path } = req.query;
    if (!path) return res.status(400).json({ error: "path query param is required" });
    const pr = await getPR(owner, repo, number);
    const content = await getFileContent(owner, repo, path, pr.head.sha);
    res.type("text/plain").send(content);
  } catch (err) {
    next(err);
  }
});

router.delete("/added-prs/:owner/:repo/:number", (req, res) => {
  const { owner, repo, number } = req.params;
  deletePR(`${owner}/${repo}`, Number(number));
  res.status(204).end();
});

File: server/app/routes/prs.py
import logging
from typing import Any

from fastapi import APIRouter, Query, Response
from pydantic import BaseModel

from app import db
from app import github
from app.concurrency import map_with_concurrency
from app.errors import AppError

logger = logging.getLogger(__name__)

router = APIRouter()


def _with_status(pr: dict[str, Any], full_db: dict[str, Any]) -> dict[str, Any]:
    return {**pr, "status": db.compute_status(pr["repo"], pr["number"], full_db)}


class AddedPRBody(BaseModel):
    url: str | None = None


@router.get("/prs")
async def list_prs() -> list[dict[str, Any]]:
    login = await github.get_authenticated_login()
    search_results = await github.search_review_requested_open_prs(login)

    assigned_keys: set[str] = set()

    async def refresh_assigned(item: dict[str, Any], _i: int) -> None:
        owned = github.repo_owner_from_url(item["repository_url"])
        owner, repo = owned["owner"], owned["repo"]
        if not owner or not repo:
            return
        try:
            full = await github.get_pr(owner, repo, item["number"])
            pr = github.normalize_pr(full, owner, repo, "assigned")
            db.upsert_pr(pr)
            assigned_keys.add(f"{owner}/{repo}#{item['number']}")
        except Exception as err:
            logger.error("Failed to refresh assigned PR %s/%s#%s: %s", owner, repo, item["number"], err)

    await map_with_concurrency(search_results, 8, refresh_assigned)

    # Refresh manually-added PRs not covered by the assigned fetch above.
    db_before = db.read_db()
    manual_entries = [
        (key, pr) for key, pr in db_before["prs"].items() if pr.get("source") == "manual" and key not in assigned_keys
    ]

    async def refresh_manual(entry: tuple[str, dict[str, Any]], _i: int) -> None:
        key, pr = entry
        owner, repo = pr["repo"].split("/")
        try:
            full = await github.get_pr(owner, repo, pr["number"])
            db.upsert_pr(github.normalize_pr(full, owner, repo, "manual"))
        except Exception as err:
            logger.error("Failed to refresh manual PR %s, keeping stale copy: %s", key, err)

    await map_with_concurrency(manual_entries, 8, refresh_manual)

    # Drop assigned PRs that are no longer open+review-requested (closed, merged, or
    # review no longer needed) — otherwise they'd linger in the db forever since only
    # currently-returned search results get upserted, never pruned.
    for key, pr in db_before["prs"].items():
        if pr.get("source") == "assigned" and key not in assigned_keys:
            db.delete_pr(pr["repo"], pr["number"])

    full_db = db.read_db()
    prs = [
        pr
        for pr in full_db["prs"].values()
        if pr.get("source") == "manual" or pr.get("state") is None or pr.get("state") == "open"
    ]
    result = [_with_status(pr, full_db) for pr in prs]
    result.sort(key=lambda pr: pr.get("updatedAt") or "", reverse=True)
    return result


@router.post("/added-prs", status_code=201)
async def add_pr(body: AddedPRBody) -> dict[str, Any]:
    if not body.url:
        raise AppError("url is required", 400)
    parsed = github.parse_pr_url(body.url)
    owner, repo, number = parsed["owner"], parsed["repo"], parsed["number"]
    full = await github.get_pr(owner, repo, number)
    pr = db.upsert_pr(github.normalize_pr(full, owner, repo, "manual"))
    return _with_status(pr, db.read_db())


@router.get("/prs/{owner}/{repo}/{number}/diff")
async def get_pr_diff(owner: str, repo: str, number: int) -> Response:
    diff = await github.get_pr_diff(owner, repo, number)
    return Response(content=diff, media_type="text/plain")


@router.get("/prs/{owner}/{repo}/{number}/file")
async def get_pr_file(owner: str, repo: str, number: int, path: str = Query(default="")) -> Response:
    if not path:
        raise AppError("path query param is required", 400)
    pr = await github.get_pr(owner, repo, number)
    content = await github.get_file_content(owner, repo, path, pr["head"]["sha"])
    return Response(content=content, media_type="text/plain")


@router.delete("/added-prs/{owner}/{repo}/{number}", status_code=204)
async def delete_added_pr(owner: str, repo: str, number: int) -> Response:
    db.delete_pr(f"{owner}/{repo}", number)
    return Response(status_code=204)


class LinkedPRBody(BaseModel):
    url: str | None = None


# Links a related PR (e.g. a frontend PR alongside the backend/library PR it depends
# on) so a review run can pull in the linked PR's diff as extra cross-PR context.
# Reference only here — the actual context-building happens in routes/reviews.py.
@router.post("/prs/{owner}/{repo}/{number}/links", status_code=201)
async def add_linked_pr(owner: str, repo: str, number: int, body: LinkedPRBody) -> dict[str, Any]:
    if not body.url:
        raise AppError("url is required", 400)
    parsed = github.parse_pr_url(body.url)
    l_owner, l_repo, l_number = parsed["owner"], parsed["repo"], parsed["number"]
    if l_owner == owner and l_repo == repo and l_number == number:
        raise AppError("A PR can't be linked to itself", 400)
    await github.get_pr(l_owner, l_repo, l_number)  # validate it actually exists before storing the link
    pr = db.add_linked_pr(f"{owner}/{repo}", number, f"{l_owner}/{l_repo}", l_number)
    if not pr:
        raise AppError("PR not found — load Assigned PRs at least once before linking", 404)
    return _with_status(pr, db.read_db())


@router.delete("/prs/{owner}/{repo}/{number}/links/{l_owner}/{l_repo}/{l_number}", status_code=204)
async def remove_linked_pr(owner: str, repo: str, number: int, l_owner: str, l_repo: str, l_number: int) -> Response:
    db.remove_linked_pr(f"{owner}/{repo}", number, f"{l_owner}/{l_repo}", l_number)
    return Response(status_code=204)
